### PR TITLE
additing models, business logic for the box, serializers

### PR DESCRIPTION
**Pull request by DimaUlasovets**

### CODE REVIEW

**Comment by DimaUlasovets** [comment]
I have a problem with black and isort. They constantly gave out "Failed" and nothing helped. What could be wrong?

**Comment by DimaUlasovets** [comment]
And should i modify all classes in views.py to mixins style ?

**Comment by vadim-denisov** [comment]
> I have a problem with black and isort. They constantly gave out "Failed" and nothing helped. What could be wrong?

Did you google it?

File: pyproject.toml [hunk 1]
```diff
@@ -11,6 +11,8 @@ djangorestframework = "^3.13.1"
 psycopg-binary = "^3.0.11"
 pre-commit = "^2.18.1"
 flake8 = "^4.0.1"
+psycopg2-binary = "^2.9.3"
+black = "^22.3.0"
```
Comment (vadim-denisov) [inline pyproject.toml]: It's dev dependency

Comment (vadim-denisov) [reply pyproject.toml]: Also not sure that we require it here, since we have it in pre-commit.

File: secret_santa/api/models.py [hunk 2]
```diff
@@ -1,3 +1,79 @@
...
 from django.db import models
 
 # Create your models here.
+
+
+STATUS_CHOICES = (
+    ("open", "Open"),
+    ("moving", "Moving"),
+    ("close", "Close"),
+)
```
Comment (vadim-denisov) [inline secret_santa/api/models.py]: Can we use enum?

File: secret_santa/api/models.py [hunk 3]
```diff
@@ -1,3 +1,79 @@
...
+    def __str__(self):
+        return self.name_box
+
+    @property
+    def users(self):
+        giftrequests = self.giftrequests.all()
+        users = []
+        for giftrequest in giftrequests:
+            users.append(giftrequest.user)
+        return users
```
Comment (vadim-denisov) [inline secret_santa/api/models.py]: `queryset to list` it usually responsibility of serializer

File: secret_santa/api/serializers.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+class GiftRequestSerializer(serializers.ModelSerializer):
+    box = BoxesSerializer(many=False)
+
+    class Meta:
+        model = GiftRequest
+        fields = "__all__"
+
+
+class UsersSerializer(serializers.ModelSerializer):
+    # gifts = serializers.SerializerMethodField()
```
Comment (vadim-denisov) [inline secret_santa/api/serializers.py]: don't store dead code

File: secret_santa/secret_santa/settings.py [hunk 5]
```diff
@@ -77,8 +77,10 @@
 
 DATABASES = {
     "default": {
-        "ENGINE": "django.db.backends.sqlite3",
-        "NAME": BASE_DIR / "db.sqlite3",
+        "ENGINE": "django.db.backends.postgresql",
+        "NAME": "secret_santa",
+        "USER": "d.ulasovets",
+        "PASSWORD": "lastochka1488",
```
Comment (vadim-denisov) [inline secret_santa/secret_santa/settings.py]: store it in env variables

File: secret_santa/api/views.py [hunk 6]
```diff
@@ -1,3 +1,175 @@
...
+class SingleReceiverSender(APIView):
+    """
+    Single receiver/sender be id
+    """
+
+    def get_single_receiver_sender(self, pk):
+        try:
+            return ReceiverSender.objects.get(pk=pk)
+        except (ValidationError, ReceiverSender.DoesNotExist):
+            raise Http404
```
Comment (vadim-denisov) [inline secret_santa/api/views.py]: use `get_or_404`

File: secret_santa/api/views.py [hunk 7]
```diff
@@ -1,3 +1,175 @@
...
+                          UsersSerializer)
+
+
+class RoutesList(APIView):
+    """
+    List all routes.
+    """
+
+    def get(self, request, format=None):
+        routes = [
```
Comment (vadim-denisov) [inline secret_santa/api/views.py:18]: You also have DELETE and POST methods to some of these endpoints

**Comment by DimaUlasovets** [comment]
i found the problem, i found the problem black and isort changed each other's changes

File: secret_santa/api/views.py (hunk 7 above)
Comment (DimaUlasovets) [reply secret_santa/api/views.py:18]: Class RoutesList only for view endpoints, i don't think that he should have post, delete method. Correct me if I'm wrong

File: secret_santa/api/models.py (hunk 2 above)
Comment (DimaUlasovets) [reply secret_santa/api/models.py:13]: Didn't understand what you mean

Comment (DimaUlasovets) [reply secret_santa/api/models.py:13]: Okay, i find what you mean, i'll try this

Comment (DimaUlasovets) [reply secret_santa/api/models.py]: As i read about Enum, does not correctly support the HTML support render of the REST Framework

File: pyproject.toml [hunk 8]
```diff
@@ -9,10 +9,14 @@ python = "^3.8"
 Django = "^4.0.4"
 djangorestframework = "^3.13.1"
 psycopg-binary = "^3.0.11"
-pre-commit = "^2.18.1"
-flake8 = "^4.0.1"
+psycopg2-binary = "^2.9.3"
+
+django-dotenv = "^1.4.2"
```
Comment (vadim-denisov) [inline pyproject.toml]:
```suggestion
psycopg2-binary = "^2.9.3"
django-dotenv = "^1.4.2"
```

File: secret_santa/api/models.py (hunk 2 above)
Comment (vadim-denisov) [reply secret_santa/api/models.py]: Can you create enum with class method `choices` which return structure like you describe above?

File: secret_santa/secret_santa/settings.py [hunk 9]
```diff
@@ -77,8 +79,10 @@
 
 DATABASES = {
     "default": {
-        "ENGINE": "django.db.backends.sqlite3",
-        "NAME": BASE_DIR / "db.sqlite3",
+        "ENGINE": "django.db.backends.postgresql",
+        "NAME": str(os.getenv("DB_NAME")),
+        "USER": str(os.getenv("DB_USER")),
+        "PASSWORD": str(os.getenv("DB_PASSWORD")),
```
Comment (vadim-denisov) [inline secret_santa/secret_santa/settings.py]: Why do we use str?

File: secret_santa/manage.py [hunk 10]
```diff
@@ -1,11 +1,19 @@
...
 
+import dotenv
+
 
 def main():
     """Run administrative tasks."""
+
+    DOT_ENV_PATH = pathlib.Path() / ".env"
+    if DOT_ENV_PATH.exists():
+        dotenv.read_dotenv(str(DOT_ENV_PATH))
```
Comment (vadim-denisov) [inline secret_santa/manage.py]: and here?

File: secret_santa/secret_santa/settings.py [hunk 11]
```diff
@@ -20,7 +21,8 @@
 # See https://docs.djangoproject.com/en/4.0/howto/deployment/checklist/
 
 # SECURITY WARNING: keep the secret key used in production secret!
-SECRET_KEY = "django-insecure-r9)!jqctn8+!v$wrzyv&ig&)gt=54q4_a9@0awpm3q16+0)j_q"
+SECRET_KEY = str(os.getenv("SECRET_KEY"))
```
Comment (vadim-denisov) [inline secret_santa/secret_santa/settings.py:24]:
```suggestion
SECRET_KEY = os.getenv("SECRET_KEY")
```

File: secret_santa/api/models.py [hunk 12]
```diff
@@ -1,3 +1,84 @@
...
+    def __str__(self):
+        return f"{self.first_name} {self.last_name}"
+
+
+class Boxes(models.Model):
+    id = models.UUIDField(default=uuid.uuid4, primary_key=True, editable=False, unique=True)
+    name_box = models.CharField(max_length=100)
+    create_date = models.DateTimeField(auto_now_add=True)
+    end_date = models.DateTimeField(blank=True, null=True)
+    status_box = models.CharField(max_length=10, choices=StatusChoices.choices(), default="open")
```
Comment (vadim-denisov) [inline secret_santa/api/models.py:35]:
```suggestion
    status_box = models.CharField(max_length=10, choices=StatusChoices.choices(), default=StatusChoices.OPEN)
```

File: secret_santa/api/models.py [hunk 13]
```diff
@@ -1,3 +1,84 @@
+import random
+import uuid
+from enum import Enum
+
 from django.db import models
 
 # Create your models here.
+
+
+class StatusChoices(Enum):
```
Comment (vadim-denisov) [inline secret_santa/api/models.py:10]: In order to compare enum and strings
```suggestion
class StatusChoices(str, Enum):
```

File: secret_santa/api/models.py [hunk 14]
```diff
@@ -1,3 +1,84 @@
...
+    def users(self):
+        giftrequests = self.giftrequests.all()
+
+        users = []
+        for giftrequest in giftrequests:
+            users.append(giftrequest.user)
+        return users
+
+    def move_bag(self):
+        if self.status_box == "moving":
```
Comment (vadim-denisov) [inline secret_santa/api/models.py:50]: use enum here

File: secret_santa/api/models.py [hunk 15]
```diff
@@ -1,3 +1,84 @@
...
+        return self.name_box
+
+    @property
+    def users(self):
+        giftrequests = self.giftrequests.all()
+
+        users = []
+        for giftrequest in giftrequests:
+            users.append(giftrequest.user)
+        return users
```
Comment (vadim-denisov) [inline secret_santa/api/models.py:47]: I think it's more method than property since it's mutable

File: secret_santa/api/models.py [hunk 16]
```diff
@@ -1,3 +1,84 @@
...
+
+    def move_bag(self):
+        if self.status_box == "moving":
+            giftrequests = self.giftrequests.all()
+            users = self.users
+            for giftrequest in giftrequests:
+                secret_user = giftrequest.user
+                while giftrequest.user == secret_user:
+                    secret_user = random.choice(users)
```
Comment (vadim-denisov) [inline secret_santa/api/models.py:56]: If we only have 1 user, will we repeat this loop infinite number of times?

Comment (DimaUlasovets) [reply secret_santa/api/models.py:56]: this function will start only  if self.status_box == "open" and len(self.users) > 2
i described it in the function below

**Pull request merged**